src/compiler/utilities.ts

```diff
@@ -2060,19 +2060,54 @@ namespace ts {
             }
             return AssignmentDeclarationKind.ObjectDefinePropertyValue;
         }
-        if (expr.operatorToken.kind !== SyntaxKind.EqualsToken ||
-            !isPropertyAccessExpression(expr.left)) {
+        if (expr.operatorToken.kind !== SyntaxKind.EqualsToken) {
             return AssignmentDeclarationKind.None;
         }
         const lhs = expr.left;
-        if (isEntityNameExpression(lhs.expression) && lhs.name.escapedText === "prototype" && isObjectLiteralExpression(getInitializerOfBinaryExpression(expr))) {
-            // F.prototype = { ... }
-            return AssignmentDeclarationKind.Prototype;
+        if (isPropertyAccessExpression(lhs) || isElementAccessExpression(lhs)) {
+            if (isEntityNameExpression(lhs.expression) && getElementOrPropertyAccessName(lhs) === "prototype" && isObjectLiteralExpression(getInitializerOfBinaryExpression(expr))) {
+                    // F.prototype = { ... }
+                    return AssignmentDeclarationKind.Prototype;
+            }
+            return getAssignmentDeclarationPropertyAccessKind(lhs);
+        }
+        return AssignmentDeclarationKind.None;
+    }
+
+    /**
+     * Does not handle signed numeric names like `a[+0]` - handling those would require handling prefix unary expressions
+     * throughout late binding handling as well, which is awkward (but ultimately probably doable if there is demand)
+     */
+    /* @internal */
+    export function getElementOrPropertyAccessArgumentExpressionOrName(node: PropertyAccessExpression | ElementAccessExpression): Identifier | StringLiteralLike | NumericLiteral | ElementAccessExpression | undefined {
+        if (isPropertyAccessExpression(node)) {
+            return node.name;
+        }
+        const arg = skipParentheses(node.argumentExpression);
+        if (isNumericLiteral(arg) || isStringLiteralLike(arg)) {
+            return arg;
+        }
+        return node;
+    }
+
+    /* @internal */
+    export function getElementOrPropertyAccessName(node: PropertyAccessExpression | ElementAccessExpression): string | undefined {
+        const name = getElementOrPropertyAccessArgumentExpressionOrName(node);
+        if (name) {
+            if (isIdentifier(name)) {
+                return idText(name);
+            }
+            if (isNumericLiteral(name)) {
+                return "" + (+name.text);

```

Comment:
🤔 I’m sure there’s a good reason for this

Reply:
Assume `name.text` is `1e1` as a string. Convert to number, get out `10`, convert back to string, have the string `"10"`. This matters for indexing, since only the canonical representation is the stringy key name.

![image](https://user-images.githubusercontent.com/2932786/65734171-9e71ac00-e086-11e9-9ff8-5f2d48941cb3.png)

Reply:
@rbuckton points out that the misleadingly named `.text` has already been canonicalized in the scanner, so this is unneeded 😛